pkg/ds/farm.go

```diff
@@ -0,0 +1,358 @@
+			if !ok {
+				return
+			}
+			if changes == nil {
+				dsf.logger.Error("Unexpected nil value received on channel")
+				return
+			}
+		}
+
+		for _, child := range dsf.children {
+			select {
+			case <-quitCh:
+				return
+			case err, ok = <-child.errCh:
+				if !ok {
+					// child error channel closed
+					return
+				}
+				dsf.logger.Fatalf("An fatal error has occurred in ds: '%v':, %v", child.ds, err)
+			default:
+			}
+		}
+		dsf.handleDSChanges(quitCh, changes)
+	}
+}
+
+func (dsf *Farm) handleDSChanges(quitCh <-chan struct{}, changes *dsstore.WatchedDaemonSets) {
+	if changes.Err != nil {
+		dsf.logger.Infof("An error has occurred while watching daemon sets: %v", changes.Err)
+	}
+
+	if len(changes.Created) > 0 {
+		dsf.logger.Infof("The following daemon sets have been created:")
+		for _, dsFields := range changes.Created {
+			dsf.logger.Infof("%v", *dsFields)
+
+			// If the daemon set contends with another daemon set, disable it
+			// if that fails, delete it, if that fails too this is a fatal error
+			dsIDContended, isContended, err := dsf.dsContends(dsFields)
+			if err != nil {
+				dsf.logger.Errorf("Error occurred when trying to check for daemon set contention: %v", err)
+				continue
+			}
+
+			if isContended {
+				dsf.logger.Errorf("Created daemon set '%s' contends with %s", dsFields.ID, dsIDContended)
+				newDS, err := dsf.dsStore.Disable(dsFields.ID)
+				if err != nil {
+					dsf.logger.Errorf("Error occurred when trying to delete daemon set: %v", err)
+					continue
+				}
+				dsf.children[newDS.ID] = dsf.spawnDaemonSet(&newDS, quitCh)
+			} else {
+				dsf.children[dsFields.ID] = dsf.spawnDaemonSet(dsFields, quitCh)
+			}
+		}
+	}
+
+	if len(changes.Updated) > 0 {
+		dsf.logger.Infof("The following daemon sets have been updated:")
+		for _, dsFields := range changes.Updated {
+			dsf.logger.Infof("%v", *dsFields)
+
+			// If the daemon set contends with another daemon set, disable it
+			// if that fails, delete it, if that fails too this is a fatal error
+			if _, ok := dsf.children[dsFields.ID]; ok {
+				dsIDContended, isContended, err := dsf.dsContends(dsFields)
+				if err != nil {
+					dsf.logger.Errorf("Error occurred when trying to check for daemon set contention: %v", err)
+					continue
+				}
+
+				if isContended {
+					dsf.logger.Errorf("Updated daemon set '%s' contends with %s", dsFields.ID, dsIDContended)
+					newDS, err := dsf.dsStore.Disable(dsFields.ID)
+					if err != nil {
+						dsf.logger.Errorf("Error occurred when trying to delete daemon set: %v", err)
+						continue
+					}
+					dsf.children[newDS.ID].updatedCh <- &newDS
+				} else {
+					dsf.children[dsFields.ID].updatedCh <- dsFields
+				}
+			}
+		}
+	}
+
+	if len(changes.Deleted) > 0 {
+		dsf.logger.Infof("The following daemon sets have been deleted:")
+		for _, dsFields := range changes.Deleted {
+			dsf.logger.Infof("%v", *dsFields)
+			if child, ok := dsf.children[dsFields.ID]; ok {
+				child.deletedCh <- dsFields
```

Comment:
also, you should probably close the channel along with child.updatedCh, and child.quitCh